components/claim/components/email_to_ldap.tsx

```diff
@@ -0,0 +1,233 @@
+
+import * as Utils from 'utils/utils';
+import {t} from 'utils/i18n.jsx';
+
+import LoginMfa from 'components/login/login_mfa';
+import LocalizedInput from 'components/localized_input/localized_input';
+
+type Props = {
+    email: string | null;
+    siteName?: string;
+    ldapLoginFieldName?: string;
+}
+
+const EmailToLDAP = ({email, siteName, ldapLoginFieldName}: Props) => {
+    const emailPasswordInput = useRef<HTMLInputElement>(null);
+    const ldapIdInput = useRef<HTMLInputElement>(null);
+    const ldapPasswordInput = useRef<HTMLInputElement>(null);
+
+    const [password, setPassword] = React.useState('');
+    const [ldapId, setLdapId] = React.useState<string | null>(null);
+    const [ldapPassword, setLdapPassword] = React.useState<string | null>(null);
+    const [passwordError, setPasswordError] = React.useState('');
+    const [ldapError, setLdapError] = React.useState('');
+    const [ldapPasswordError, setLdapPasswordError] = React.useState('');
+    const [serverError, setServerError] = React.useState('');
+    const [showMfa, setShowMfa] = React.useState(false);
+
+    const preSubmit = (e: React.FormEvent) => {
+        e.preventDefault();
+
+        const password = emailPasswordInput.current && emailPasswordInput.current.value;
+        if (!password) {
+            setPasswordError(Utils.localizeMessage('claim.email_to_ldap.pwdError', 'Please enter your password.'));
+            return;
+        }
+
+        const ldapId = ldapIdInput.current && ldapIdInput.current.value.trim();
+        if (!ldapId) {
+            setLdapError(Utils.localizeMessage('claim.email_to_ldap.ldapIdError', 'Please enter your AD/LDAP ID.'));
+            return;
+        }
+
+        const ldapPassword = ldapPasswordInput.current && ldapPasswordInput.current.value;
+        if (!ldapPassword) {
+            setLdapPasswordError(Utils.localizeMessage('claim.email_to_ldap.ldapPasswordError', 'Please enter your AD/LDAP password.'));
+            return;
+        }
+
+        setPassword(password);
+        setLdapId(ldapId);
+        setLdapPassword(ldapPassword);
+
+        if (email) {
+            submit(email, password, '', ldapId, ldapPassword);
+        }
+    };
+
+    const submit = (loginIdParam: string, passwordParam: string, token: string, ldapIdParam?: string, ldapPasswordParam?: string) => {
+        emailToLdap(
+            loginIdParam,
+            passwordParam,
+            token,
+            ldapIdParam || ldapId,
+            ldapPasswordParam || ldapPassword,
+            (data: AuthChangeResponse) => {
+                if (data.follow_link) {
+                    window.location.href = data.follow_link;
+                }
+            },
+            (err: {server_error_id: string; id: string; message: string}) => {
+                if (!showMfa && err.server_error_id === 'mfa.validate_token.authenticate.app_error') {
+                    setShowMfa(true);
+                } else {
+                    switch (err.id) {
```

Comment:
I am not sure I got your point on this using `const` or `enum`. Because in `email_to_ldap.tsx` we are using `err.id` only here

Reply:
yes, but we use the results (e.g. `'mfa.validate_token.authenticate.app_error'`) in other, at least one, places, so we could provide an enum like this:

```
export enum AppErrors {
    ENT_LDAP_LOGIN_USER_NOT_REGISTERED = 'ent.ldap.do_login.user_not_registered.app_error',
    ENT_LDAP_LOGIN_USER_FILTERED = 'ent.ldap.do_login.user_filtered.app_error',
    ...
}
```

and use it like this:

```
switch (err.id) {
case AppErrors.ENT_LDAP_LOGIN_USER_NOT_REGISTERED:
case AppErrors.ENT_LDAP_LOGIN_USER_FILTERED:
...
```

That way we create a central and single point of truth in case we use it elsewhere already or in the future.
I think it might be well placed in the `constants.jsx` file, but since that is not converted to typescript yet we should build a standard `const` (object) ...

```
export const AppErrors = {
    ENT_LDAP_LOGIN_USER_NOT_REGISTERED: 'ent.ldap.do_login.user_not_registered.app_error',
    ENT_LDAP_LOGIN_USER_FILTERED: 'ent.ldap.do_login.user_filtered.app_error',
    ...
}
```

... and convert it to an `enum` later on. Maybe put a comment onto that once you have build it so we can revisit it whenever we get the chance to convert that file to typescript.

Feel free to reach out if you have any additional questions. 💪🏼

Reply:
Ahh, now I see your point which totally makes sense. I converted it to use enum in `utils.constants.tsx`. I am not sure which `constants.jsx` file you were referring

Reply:
Oh, I was under the impression that the constants file is still a `jsx` for some reason